flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobSubmitHandler.java

```diff
@@ -66,6 +67,9 @@ public JobSubmitHandler(
 		}
 
 		return gateway.submitJob(jobGraph, timeout)
-			.thenApply(ack -> new JobSubmitResponseBody("/jobs/" + jobGraph.getJobID()));
+			.thenApply(ack -> new JobSubmitResponseBody("/jobs/" + jobGraph.getJobID()))
+			.exceptionally(exception -> {
+				throw new CompletionException(new RestHandlerException("Job submission failed.", HttpResponseStatus.INTERNAL_SERVER_ERROR, exception));
```

Comment:
Maybe we could add the `exception.getMessage` to the `message` of the `RestHandlerException`. Otherwise the user will only see `"Job submission failed."` in the `ErrorResponseBody`. With the change it could be `"Job submission failed: Failure cause"`

Reply:
I'm not quite fond of the idea. As alluded in the PR description and JIRA I agree that the existing error message isn't _helpful_, yet better than the current state.

I rather like that so far the REST API has control over the error messages. This ensures that the user only sees messages that were actually meant for him.

In contrast, exception messages are pretty much arbitrary. They may change at will, the audience isn't defined (user vs dev), may only helpful if the fully stack trace is present, often don't have any message at all (see usages of `Preconditions`, or NPEs) and typically only describe what went wrong, not why, how to fix it or if it even was a user-error. Given that this would break down the barrier between internal/user-facing messages you obviously also run into cases where users have _no idea_ what the message even means. Finally you end up with mismatches between the error message and error code.

To me the underlying issue is that `submitJob` funnels all manner of exceptions into a `FlinkException/JobSubmissionException` that we can't do much with. Neither can we categorize them in any way, nor distinguish between who's responsible (user vs Flink) nor when in the process the failure occurred.
Without diving into the implementation you don't _even know which exceptions are thrown_, but i suppose this is a general issue of `CompletableFutures`.

Reply:
I see your point. I'm just wondering whether some bits of context wouldn't be helpful on the client side when using the CLI. So for example if the job was misconfigured or if it was already submitted to the cluster in HA mode, then it would be helpful for the user to know.

Reply:
well, there's no doubt that it _could_ be helpful; my point is that it can be _harmful_ if not done properly.

The `submitJob` should either provide the `JobSubmitHandler` with means to detect these exceptions and create adequate responses, or explicitly throw exceptions with messages that we can safely pass on to users.

That said, I do not know how to do either of these things in a good way. 😞 

For completeness sake, here are ideas that came to mind:

## 1
Introduce a special `FlinkUserFacingException` that we "trust" to contain a good error message.

Con: This provides little additional safety and will never provide proper HTTP response code.

## 2
Introduce dedicated exceptions for the scenarios that you listed and explicitly look for them in the `exceptionally` block, i.e
```
.exceptionally(exception -> {
	if (exception instanceof JobAlreadyExistsException) {
		throw new CompletionException(new RestHandlerException("Job already exists.", HttpResponseStatus.BAD_REQUEST, exception));
	} else {
		throw new CompletionException(new RestHandlerException("Job submission failed.", HttpResponseStatus.INTERNAL_SERVER_ERROR, exception));
	}
}
```

Con: Obviously, this approach is inherently flawed as there is no guarantee that a given exception can be thrown; we would have to manually keep it in sync with the actual implementation because `CompletableFuture` throw a wrench into sane exception handling. 😡 

## 3
Encode possible user-facing exceptions in the return value of `submitJob`, i.e. return a `AckOrException`
```
public class AckOrException {
	// holds exception, could also be a series of nullable fields
	private final SuperEither<ExceptionA, ExceptionB, ExceptionC> exception; 
	...
	public void throwIfError() throws ExceptionA, ExceptionB, ExceptionC;
}
```
Con: Relies on users to call `throwIfError` and introduces an entirely separate channel for passing errors, but it would allow exception matching.

Reply:
Do note that this discussion isn't really blocking the PR from being merged as it would effectively be an extension of it.

Reply:
I would be in favor of approach 3 because we are doing something similar for the `JobExecutionResult`/`JobResult`. We could then throw the exception on the `RestClusterClient`. And I also agree that this is something we can add as a follow up. Can you please create a JIRA issue for this @zentol.